Blazor with Graph API

guardrex opened the pull request:
Fixes #20144
Fixes #19503

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/graph-api?view=aspnetcore-3.1&branch=pr-en-us-20192)

guardrex commented:
@jmprieur ... This is *very* prototyped at this point. I don't think you should look at the **_whole PR_** because there are a lot of updates that aren't germane to our discussion on the issue ... we can handle the rest of the updates here on our side. The thing here that I need the most help with is the server API *daemon app* approach.

**Scenario:** A hosted Blazor solution where the client WASM app **ONLY** has access to the server API (i.e., API controllers). The server API app has **Application** permission (scope) access to Graph API. When the server API app validates the user's access token, it uses the user's `OID` claim to make a Graph API call to `memberOf` to obtain the user's Azure security groups and roles, which it then uses to establish claims for the user. These claims permit fine-grained access to API controllers via preconfigured auth policies. In this way, the client never has any kind of *on-behalf-of* access (`access_as_user`) to Graph API's `memberOf` request, which requires (minimally) `Directory.Read.All` permission.

What I've done **_works_** ... but it will need more refactoring. It is prototyped, and the whole thing only just started working this morning. I need to sleep 💤🛏️ on it and hit it again tomorrow (Tuesday) morning and take your feedback into account.

You're going to see that I've gone to considerable trouble of creating a service to obtain the `GraphServiceClient`. The service client is created when the app starts via an `IHostedService`. I'm not sure if that's **_overkill_** here. Originally, I did just have ...

* `Startup.ConfigureServices` perform `ConfidentialClientApplicationBuilder.Create`.
* The `OnTokenValidated` method:
  * Call `AcquireTokenForClient`.
  * Call `new GraphServiceClient`.

... and then make the request with `graphClient.Users[oid].MemberOf.Request().GetAsync()`. However this afternoon, I felt more like going the direction on the PR ... split those out into a service. Let the `OnTokenValidated` method use the service to get the Graph service client. I'm a bit worried at the moment if the access token will timeout at some point and the Graph client then fails. Let's talk about that if that's what is going to happen here.

Otherwise, is there anything particularly good about my splitting out into a service and obtaining the graph client on startup (e.g., no delay on first API request)? Anything particularly bad about it? ... other than creating all of this **_extra code_**? :smile:

On the diff ... scroll down and open ...

**aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md**

The section starts at **Line 508** with the section heading ...

**Authorize server API access for user-defined groups and Administrator Roles**

You'll need to expand the code and text for bits that the diff has closed by default. The whole section runs from Line 508 down to Line 859 where you see the moniker range end for 5.0.

Alternatively, I'm going to add a link to the internal review topic in a minute. You could flip that to 5.0 and see it that way. Stand-by ... I'll add that right now.

guardrex commented:
@jmprieur I placed the code directly into the comment above :point_up: to make it easy for you to run an eye over it.

### **_However ..._**

WRT the new approaches on https://github.com/AzureAD/microsoft-identity-web/pull/705 ... I'm working with `Microsoft.Graph` 3.18.0 and `Microsoft.Identity.Client` 4.18.0, but I see that `Microsoft.Identity.Client` 4.21.1 is out. Does that package contain the new API for the latest daemon approach from the PR? If so, then I need to update my process here and then ping u back next week. If I can't use the new daemon approach yet, then I'll open a new issue, and we should proceed with what I'm working on here because I need to get these updates in for 5.0 GA.

Tell me one thing tho if I'm going to update to the latest daemon app approach now: Do I need to consider if the Graph access token ever expires in a daemon app? I think that I don't need to worry about that, but I just want to check with you on it.

jmprieur commented:
@guardrex : we released Microsoft.Identity.Web 1.2.0 today.
It does have the new daemon approach, which is explained here: https://github.com/AzureAD/microsoft-identity-web/wiki/1.2.0#you-can-now-specify-scopes-and-app-permissions-for-graphserviceclient

If the token expires, Microsoft.Identity.Web (via MSAL.NET) will renew it. So no worries about that

guardrex commented:
@jmprieur ... *Awesome!* ... Thanks ... ok ... I'll rework that bit then and ping u back next week (by Tuesday morning). I'll put the approach into a **_comment_** here to make it quick for you to inspect.

cc: @captainsafia ... TL;DR ... The final bit for 5.0 WASM security updates here should be done Monday/Tuesday. @jmprieur will :eye: a piece of this first to make sure that I'm doing good things with the hosted server API AAD group policies+Graph SDK/API piece ... the *very last bit* that needs work here 😅. After I have that resolved (and make a few other nit updates), I'll ping u. This PR has all of the remaining 5.0 WASM work AFAIK. ..... In fact AFAIK, this PR will wrap up **_ALL_** of the major Blazor 5.0 updates :tada:🎈🍾.

jmprieur commented:
@guardrex : what I don't fully understand is why you don't start from a project generated with `dotnet new blazorserver --auth SingleOrg --calls-graph` ? (these are your templates)

Then for the groups, did you manage the overflow? 
https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/8189f7ccf02b59de67e5d1ffe8404000830de510/5-WebApp-AuthZ/5-2-Groups/Startup.cs#L51

guardrex commented:
@jmprieur ... This is a special case scenario for **WASM hosted**-generated apps. Also, we don't want an on-behalf-of the user call to Graph API where the user/client app has `Directory.Read.All` permission (scope) in the server API app. We only want the user/client to have a permission/scope **_for the server API_**. The server API has the permission/scope for Graph API and `Directory.Read.All` permission. This is a special case scenario and not in the main Blazor WASM security docs. It only appears in the AAD groups and roles topic and only for apps that need to set up auth policies for controller endpoints. It's an edge case for sure, but I wanted to cover it for completeness here with the rest of our AAD groups and roles coverage. We had a reader ask for it, and I thought it was a good idea to have it.

**_Yes!_** on handling the presence of a `hasgroups` claim ... overflow is under control here and demonstrated in our code examples.

guardrex commented:
@captainsafia ... I think we're ready here. I've looked over the Graph API/SDK docs again this morning. I reverted the code on this PR back to making the `AcquireTokenForClient` request in `Startup.ConfigureServices` because the token granted to a daemon app does expire and must be refreshed. According to Azure docs, an **_application_** (not user) token cache works behind-the-scenes obtaining a new token if needed and storing them for future use. It all comes pretty much straight out of [the Azure docs for the daemon app](https://docs.microsoft.com/azure/active-directory/develop/scenario-daemon-overview). I reiterate that this is only for the AAD groups and roles topic and only for a server API (web API) that needs to independently set up auth policies for controller endpoints without on-behalf-of user requests to Graph API and without additional permission/scopes for users granted to the client app. The user only ever has a permission/scope for the server API in this scenario.

WRT [Javier's Graph code](https://github.com/dotnet/AspNetCore.Docs/issues/20144), that's on the PR. That code has been placed into a new Graph API topic. That code is also used in the AAD groups and roles for the Graph SDK approach in the client-side app.

I've added the VS NOTE that DR recommends for 5.0, which we can remove later when VS releases (I'll open a tracking issue so that I don't forget). The wording of this NOTE probably needs a bit of help. You'll see these NOTEs at the tops of the Standalone and Hosted AAD topics. He said 'organization' account, so I assumed that he means multi-tenant apps created by VS are ok (i.e., B2C). If not tho, let me know :ear:, and I'll add the NOTE to the Standalone and Hosted AAD **B2C** topics.

AFAIK, we're sticking to the plan to have a final Javier (+probably others) review everything prior to GA from the live preview topics.

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
```diff
@@ -72,6 +72,172 @@ public class CustomUserAccount : RemoteUserAccount
+                    // Optional: Log the error
+
+#if DEBUG
+                    Console.WriteLine(
```
```suggestion
                   logger.LogWarning(
```

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
```diff
@@ -72,6 +72,172 @@ public class CustomUserAccount : RemoteUserAccount
+
+The preceding code doesn't obtain transitive memberships. If the app requires direct and transitive group membership claims, change the code:
+
+* For the code line:
```
```suggestion
* Replace the following line:
```

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
````diff
@@ -72,6 +72,172 @@ public class CustomUserAccount : RemoteUserAccount
+  IUserMemberOfCollectionWithReferencesPage groupsAndAzureRoles = null;
+  ```
+
+  Replace the preceding line with:
````
```suggestion
  With:
```

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
````diff
@@ -72,6 +72,172 @@ public class CustomUserAccount : RemoteUserAccount
+  IUserTransitiveMemberOfCollectionWithReferencesPage groupsAndAzureRoles = null;
+  ```
+
+* For the code line:
````
```suggestion
* Replace the line:
```

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
````diff
@@ -72,6 +72,172 @@ public class CustomUserAccount : RemoteUserAccount
+  groupsAndAzureRoles = await graphClient.Users[oid].MemberOf.Request().GetAsync();
+  ```
+
+  Replace the preceding line with:
````
```suggestion
  With:
```

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
```diff
@@ -237,6 +403,11 @@ There's no need to provide code to remove the original `groups` claim, if presen
+using Microsoft.AspNetCore.Components.WebAssembly.Authentication;
+using Microsoft.Extensions.Configuration;
+
+...
```
What does the "..." imply here?

guardrex replied:
More code that's skipped. We use the convention all the time. Readers seem ok with it.

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
```diff
@@ -382,6 +555,40 @@ For example:
+>   }
+> }
+>
+> In the preceding configuration, the end of the `Audience` value does **not** include the default scope `/{DEFAULT SCOPE}`.
```
Maybe move this above the example for cohesion. And structure the example so that it shows the sample value `https://contoso.onmicrosoft.com/41451fa7-82d9-4673-8fa5-69eff5a761fd` and highlights which parts are the tenant and app id in a table below?

captainsafia commented on aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md:
```diff
@@ -414,12 +621,166 @@ public class BillingDataController : ControllerBase
+
+### Services
+
+In the *Server* app's `Startup.ConfigureServices` method, additional namespaces are required for the code in the `Startup` class of the *Server* app. The following set of `using` statements includes the required namespaces for the code that follows in this section and the default namespaces required by a server API app created from the hosted Blazor project template:
```
Can we figure out a way to highlight what the additional and default namespaces are?

guardrex replied:
Yes ... I'll do that on the next commit.

captainsafia reviewed (commented):
Not done with the review but figured I'd post what I have at the moment. Will circle back with more feedback later.

guardrex commented:
Thanks for your help @jmprieur ... I think we're ready to move forward here. This content and these examples will be reviewed again by Javier (and others) later but prior to GA, so we'll fix them up further from here.

btw --- I did notice one thing in passing when I was looking at the Groups sample that you cross linked. The example has app settings for the group memberships that say ...

> Enter the objectID for GroupAdmin group copied from Azure Portal
> Enter the objectID for GroupMember group copied from Azure Portal

https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/8189f7ccf02b59de67e5d1ffe8404000830de510/5-WebApp-AuthZ/5-2-Groups/appsettings.json

Focusing on the remark "**_copied from Azure Portal_**" from the file, I don't think those are in the portal, nor are they published anywhere in Azure docs that I can find. I've tried a few times to track them down, including speaking with Kyle Marsh on https://github.com/dotnet/AspNetCore.Docs/issues/18924. I never found them, so we published them over here at ...

https://docs.microsoft.com/aspnet/core/blazor/security/webassembly/azure-active-directory-groups-and-roles#aad-administrator-role-object-ids

I don't think this is the best spot for these. I think they should be in the Azure docs, but I'm not sure how to address it further. Let me know if you think I should take some additional action on it.

guardrex commented:
> Will circle back with more feedback later.

I'll wait until ur done and then resolve everything at once.

guardrex commented:
@captainsafia ... Looks good thus far. I'm going to go ahead and merge now and take live to get us beyond *half-baked 5.0 preview topics*. I intend to go over it all again by the end of this week and probably make some further nit updates.

FYIs on the latest updates:

* I addressed your feedback, but I couldn't always do it in the way that you suggested. For example, we can't use directional terms in topics (e.g., "before," "after," etc.), and we mostly avoid future tense. I made changes in most cases along the lines of what u suggested.
* Some of your suggestions just led to merely killing off 🔪 non-essential language, which worked out well to focus the text.
* I caught myself writing "permission (scope)" or "permissions (scopes)" too much. This happened because the Azure portal overloads the concept with both terms. I refactored the PR a bit and mostly just go with "scope" and "scopes." It streamlines the text and is a bit more technically specific. I keep the word "permission" in a few important Azure portal-focused text.

I have one ❓ on the code that bugs me ...

In the new Graph API topic, `AddGraphClient` conveniently takes an array of scopes ...


```csharp
public static IServiceCollection AddGraphClient(
    this IServiceCollection services, params string[] scopes)
```

```csharp
builder.Services.AddGraphClient("https://graph.microsoft.com/User.Read");
```

However, `AuthenticateRequestAsync` has a hard-coded scope ...

```csharp
new AccessTokenRequestOptions()
{
    Scopes = new[] { "https://graph.microsoft.com/User.Read" }
});
```

I wish it could be refactored so that `AccessTokenRequestOptions` receives the scope array passed to `AddGraphClient`. I'm not confident that it's possible/reasonable given that Javier didn't just set it up that way in the first place. Is it possible/reasonable? If so, what would that look like?
